File: app/src/main/java/com/example/olesya/rxjavatest/view/CardActivity.java
package com.example.olesya.rxjavatest.view;

import android.app.AlertDialog;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.View;
import android.widget.NumberPicker;

import com.example.olesya.rxjavatest.Card;
import com.example.olesya.rxjavatest.ClassModels.ServiceHolderActivity;
import com.example.olesya.rxjavatest.Client;
import com.example.olesya.rxjavatest.ItemTouchCallback;
import com.example.olesya.rxjavatest.R;
import com.example.olesya.rxjavatest.Utils;
import com.example.olesya.rxjavatest.adapter.CardPagerAdapter;
import com.example.olesya.rxjavatest.databinding.ActivityCardImaginariumBinding;
import com.example.olesya.rxjavatest.interfaces.ClientCallback;

import java.net.InetAddress;
import java.util.ArrayList;

import jp.wasabeef.recyclerview.animators.FadeInDownAnimator;

public class CardActivity extends ServiceHolderActivity implements ClientCallback {

    private CardPagerAdapter mAdapter;
    private ItemTouchCallback itemTouchCallback;
    private ActivityCardImaginariumBinding mBinding;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        String username = null;
        if (getIntent().getExtras() != null)
            username = getIntent().getExtras().getString(Utils.CLIENT_CONFIG.USERNAME);
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_card_imaginarium);
        InetAddress screenAddress = (InetAddress) getIntent()
                .getSerializableExtra(Utils.CLIENT_CONFIG.HOST_CONFIG);
        startClientService(screenAddress, username);
        mBinding.buttonSend.setOnClickListener(v -> {
            Client client = (Client) getService();
            String message = mBinding.testMsg.getText().toString();
            client.onUserAction(message);
        });

        initCardPager();
    }

    @Override
    public void setCallbacks() {
        ((Client) mService).setCallbacks(this);
//        ((Client) mService).startHandlingEvents();
    }

    protected void startClientService(InetAddress screenAddress, String username) {
        mServiceIntent = new Intent(this, Client.class);
        mServiceIntent.putExtra(Utils.ACTION_SERVER_SERVICE, false);
        mServiceIntent.putExtra(Utils.CLIENT_CONFIG.HOST_CONFIG, screenAddress);
        mServiceIntent.putExtra(Utils.CLIENT_CONFIG.USERNAME, username);
        startService(mServiceIntent);
        bindService(this);
    }

    private void initCardPager() {
        mBinding.cardRv.setHasFixedSize(true);
        mBinding.cardRv.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false));
        mBinding.cardRv.setItemAnimator(new FadeInDownAnimator());
        mBinding.cardRv.getItemAnimator().setAddDuration(100);

        mAdapter = new CardPagerAdapter(new ArrayList<>());
        mAdapter.setClientCallback(this);
        mBinding.cardRv.setAdapter(mAdapter);
        itemTouchCallback = new ItemTouchCallback(mAdapter);
        ItemTouchHelper touchHelper = new ItemTouchHelper(itemTouchCallback);
        touchHelper.attachToRecyclerView(mBinding.cardRv);
    }

    @Override
    public void addCardCallback(String cardUrl) {
        runOnUiThread(() -> {
            mAdapter.addItem(new Card(cardUrl));
            mBinding.cardRv.scrollToPosition(0);
        });
    }

    @Override
    public void onMainTurnEvent() {
        runOnUiThread(() -> mBinding.title.setText(R.string.choose_topic_and_card));
        itemTouchCallback.setSwipeEnabled(true);
        mAdapter.setMainCaller(true);
    }

    @Override
    public void onMainFinishTurnEvent(String card) {
        runOnUiThread(() -> mBinding.title.setText(R.string.waiting_others));
        serverMessage.postValue(Utils.CLIENT_COMMANDS.CLIENT_MAIN_FINISHED + Utils.DELIM + card);
        itemTouchCallback.setSwipeEnabled(false);
    }

    @Override
    public void onMainStopRoundEvent() {
        runOnUiThread(() ->
                new AlertDialog.Builder(this, AlertDialog.THEME_DEVICE_DEFAULT_DARK)
                        .setTitle(R.string.finish_round)
                        .setMessage(R.string.finish_round)
                        .setNeutralButton(R.string.OK,
                                (dialog, which) -> serverMessage.postValue(Utils.CLIENT_COMMANDS.CLIENT_MAIN_STOP_FINISHED))
                        .show());
    }

    private void initChooseDialog(int maxNum) {
        View dialogView = getLayoutInflater().inflate(R.layout.dialog_number_picker, null);
        final NumberPicker np = dialogView.findViewById(R.id.numberPicker);
        np.setMaxValue(maxNum);
        np.setMinValue(1);
        np.setValue(1);
        np.setWrapSelectorWheel(true);
        new AlertDialog.Builder(this)
                .setTitle(R.string.most_suitable_card)
                .setView(dialogView)
                .setCancelable(false)
                .setNeutralButton(R.string.OK, (dialog, which) ->
                        serverMessage.postValue(Utils.CLIENT_COMMANDS.CLIENT_USER_CHOOSE_FINISHED
                                + Utils.DELIM
                                + np.getValue())).show();
    }

    @Override
    public void onUserChooseEvent(int playersNum) {
        runOnUiThread(() -> initChooseDialog(playersNum));
    }

    @Override
    public void onUserTurnEvent() {
        runOnUiThread(() -> mBinding.title.setText(R.string.user_turn));
        itemTouchCallback.setSwipeEnabled(true);
        mAdapter.setMainCaller(false);
    }

    @Override
    public void onUserFinishTurnEvent(String card) {
        serverMessage.postValue(Utils.CLIENT_COMMANDS.CLIENT_USER_FINISHED + Utils.DELIM + card);
        runOnUiThread(() -> {
            mBinding.title.setText(R.string.finish_turn);
            itemTouchCallback.setSwipeEnabled(false);
        });

    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        getService().onDestroy();
    }
}

File: app/src/main/java/com/example/olesya/boardgames/ui/view/CardActivity.kt
package com.example.olesya.boardgames.ui.view

import android.app.AlertDialog
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import android.content.Intent
import androidx.databinding.DataBindingUtil
import android.os.Bundle
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.ItemTouchHelper
import android.widget.NumberPicker
import com.example.olesya.boardgames.Commands
import com.example.olesya.boardgames.Commands.DELIM
import com.example.olesya.boardgames.R
import com.example.olesya.boardgames.Utils
import com.example.olesya.boardgames.adapter.CardDiffUtil
import com.example.olesya.boardgames.adapter.CardPagerAdapter
import com.example.olesya.boardgames.connection.client.ClientService
import com.example.olesya.boardgames.connection.common.ServiceHolderActivity
import com.example.olesya.boardgames.databinding.ActivityCardImaginariumBinding
import com.example.olesya.boardgames.ui.ItemTouchCallback
import com.example.olesya.boardgames.ui.viewmodel.CardViewModel
import jp.wasabeef.recyclerview.animators.FadeInDownAnimator
import java.io.Serializable

class CardActivity : ServiceHolderActivity() {

    private lateinit var mBinding: ActivityCardImaginariumBinding
    private val mAdapter: CardPagerAdapter = CardPagerAdapter()
    private val itemTouchCallback: ItemTouchCallback = ItemTouchCallback()
    lateinit var mViewModel: CardViewModel


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_card_imaginarium)
        mViewModel = ViewModelProviders.of(this).get(CardViewModel::class.java)
        initClientService(intent)
        initListeners()
        initCardPager()
    }

    private fun initCardPager() {
        mBinding.cardRv.setHasFixedSize(true)
        mBinding.cardRv.layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        mBinding.cardRv.itemAnimator = FadeInDownAnimator()
        (mBinding.cardRv.itemAnimator as FadeInDownAnimator).addDuration = 100
//        mAdapter.setClientCallback(this);
        mBinding.cardRv.adapter = mAdapter
        val touchHelper = ItemTouchHelper(itemTouchCallback)
        touchHelper.attachToRecyclerView(mBinding.cardRv)
    }

    private fun initListeners() {
        itemTouchCallback.pickedData.observe(this, Observer {
            if (it != null) {
                val card = mAdapter.dataset[it]
                mAdapter.remove(it)
                mViewModel.onPicked(it)
                serverMessage.postValue(Commands.CLIENT_COMMANDS.CLIENT_TURN + DELIM + card.img)
            }
        })

        mBinding.buttonSend.setOnClickListener {
            val message = mBinding.testMsg.text.toString()
            (service as ClientService).onUserAction(message)
        }

        mViewModel.player.cards.observe(this, Observer { cards ->
            //scroll to position
            if (cards != null) {
                val diffResult = DiffUtil.calculateDiff(CardDiffUtil(mAdapter.dataset, cards))
                mAdapter.setData(cards)
                diffResult.dispatchUpdatesTo(mAdapter)
            }
        })

        mViewModel.picking.observe(this, Observer { value ->
            value?.let { itemTouchCallback.isSwipeEnabled = value }
        })

        mViewModel.message.observe(this, Observer {
            run {
                if (it == null)
                    return@run

                Utils.showSnackbar(mBinding.root, it)
            }
        })

        mViewModel.choosing.observe(this, Observer {
            if (it != true)
                return@Observer

            val dialogView = layoutInflater.inflate(R.layout.dialog_number_picker, null)
            val np = dialogView.findViewById<NumberPicker>(R.id.numberPicker).apply {
                maxValue = mViewModel.playersNumber
                minValue = 1
                value = 1
                wrapSelectorWheel = true
            }

            AlertDialog.Builder(this)
                    .setTitle(R.string.most_suitable_card)
                    .setView(dialogView)
                    .setCancelable(false)
                    .setNeutralButton(R.string.OK) { _, _ ->
                        serverMessage.postValue(Commands.CLIENT_COMMANDS.CLIENT_CHOOSE
                                + Commands.DELIM
                                + np.value)
                        mViewModel.choosing.value = false
                    }.show()
        })
    }

    private fun initClientService(intent: Intent?) {
        if (intent == null || intent.extras == null)
            return

        val username = intent.extras.getString(Commands.CLIENT_CONFIG.USERNAME)
        val screenAddress = intent.getSerializableExtra(Commands.CLIENT_CONFIG.HOST_CONFIG)
        startClientService(screenAddress, username)
    }

    private fun startClientService(screenAddress: Serializable?, username: String?) {
        mServiceIntent = Intent(this, ClientService::class.java)
        mServiceIntent.putExtra(Commands.ACTION_SERVER_SERVICE, false)
        mServiceIntent.putExtra(Commands.CLIENT_CONFIG.HOST_CONFIG, screenAddress)
        mServiceIntent.putExtra(Commands.CLIENT_CONFIG.USERNAME, username)
        startService(mServiceIntent)
        bindService(this)
    }

    override fun setCallbacks() {
        (mService as ClientService).callback = mViewModel
        (mService as ClientService).start()
    }

    override fun onDestroy() {
        super.onDestroy()
        stopService(mServiceIntent)
    }
}